example/broker.cpp

```diff
@@ -51,16 +77,18 @@ void run_broker(boost::program_options::variables_map const& vm)
 #endif // defined(MQTT_USE_WS)
 
 #if defined(MQTT_USE_TLS)
-        std::unique_ptr<test_server_tls> s_tls;
+        std::shared_ptr<test_server_tls> s_tls;
```

Comment:
Why shared_ptr is needed? I think that unique_ptr is enough.

Reply:
I use shared ptr to bind to lambda for timer callback. But maybe use weak reference?

Reply:
I think that you can use unique_ptr

Reply:
unique_ptr can move capture. I can't see where is the point that we can't use unique_ptr, so far.

Reply:
yes, but then the reference in the 'run_broker' is invalidated. Maybe at some point you want to add a shutdown of the broker (shutdown all servers), or maybe request the state of all servers (for example get diagnostics of number of connected clients). I don't think you want to move the pointer into the lambda of the reload timer. The reload timer is not the owner of the server, the broker is.

Reply:
Just keep unique_ptr and pass the reference of pointee to reload_ctx.

Reply:
I noticed that https://github.com/redboltz/mqtt_cpp/commit/827ffea8117a8c0ea97d87aab7fc1e3019e53f28#diff-a980e7b60a024a0aa2c4bccacbeb0369bb88da50f872f356a1a30095c941cc03 is not good.
I didn't notice at that time.

https://github.com/redboltz/mqtt_cpp/commit/827ffea8117a8c0ea97d87aab7fc1e3019e53f28#diff-a980e7b60a024a0aa2c4bccacbeb0369bb88da50f872f356a1a30095c941cc03R41
shouldn't be a unique_ptr.
Just locate on the stack.
If nullable is required, you should use optional, not unique_ptr.

I mean

```cpp
        MQTT_NS::optional<test_server_no_tls> s;
        if (vm.count("tcp.port")) {
            s.emplace(ioc, b, vm["tcp.port"].as<uint16_t>());
        }
```

Reply:
I added https://github.com/redboltz/mqtt_cpp/wiki/Coding-Rules#choosing-variable-type-policy